src/adsr.rs

```diff
@@ -0,0 +1,103 @@
+//! Attack-Decay-Sustain-Release Envelopes
+//!
+//! These envelopes are built upon the
+//! [`envelope()`](https://docs.rs/fundsp/0.9.0/fundsp/prelude/fn.envelope.html) function to
+//! control volume over time.
+//!
+//! When a sound begins, its volume increases from zero to one in a time interval called the
+//! "Attack". It then decreases from 1.0 to the "Sustain" volume in a time interval called the
+//! "Decay". It remains at the "Sustain" level for a while, after which it decreases from the
+//! "Sustain" level to 0.0 in a time interval called the "Release".
+//!
+//! Two envelopes are available: `adsr_live()` and `adsr_fixed()`.
+//! * The `adsr_live()` function does not have a predetermined "Sustain" time interval. Instead,
+//! it awaits a `SoundMsg::Release` message by way of its `note_m`
+//! [`AtomicCell`](https://docs.rs/crossbeam/latest/crossbeam/atomic/struct.AtomicCell.html)
+//! parameter indicating when to begin the release. It signals completion of the release by placing
+//! a `SoundMsg::Finished` message in the `note_m` parameter.
+//! * The 'adsr_fixed()` function works similarly, except that it does have a predetermined
+//! "Sustain" time interval. It will release early if it receives a `SoundMsg::Release` message.
+//! It also signals completion with a `SoundMsg::Finished` message.
+//!
+//! The example `live_adsr.rs` is a fully functional demonstration of `adsr_live()`. It will
+//! listen to messages from the first connected MIDI input device it finds, and play the
+//! corresponding pitches with the volume moderated by an `adsr_live()` envelope.
+
+
+use std::fmt::Debug;
+use std::sync::Arc;
+use crossbeam::atomic::AtomicCell;
+use super::Float;
+use super::prelude::{An, Envelope, lerp, lfo};
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum SoundMsg {
+    Play, Release, Finished
+}
+
+/// Returns a volume-modulating ADSR envelope. Release occurs when a `SoundMsg::Release` message
+/// is stored in `note_m`.
+pub fn adsr_live<F:Float>(attack: F, decay: F, sustain: F, release: F, note_m: Arc<AtomicCell<SoundMsg>>) -> An<Envelope<F, F, impl Fn(F)->F + Sized + Clone, F>> {
+    adsr(attack, decay, sustain, release, None, note_m)
+}
+
+/// Returns a volume-modulating ADSR envelope. Release occurs when a `SoundMsg::Release` message
+/// is stored in `note_m`, or when `sustain_time` has elapsed, whichever comes first.
+pub fn adsr_fixed<F:Float>(attack: F, decay: F, sustain_time: F, sustain_level: F, release: F, note_m: Arc<AtomicCell<SoundMsg>>) -> An<Envelope<F, F, impl Fn(F)->F + Sized + Clone, F>> {
+    adsr(attack, decay, sustain_level, release, Some(attack + decay + sustain_time), note_m)
+}
```

Comment:
Since this is going to in public API, I would rather users not have to add `crossbeam` to their dependencies just to use `AtomicCell`.

Reply:
I am uncertain as to how to use `AudioNode::set` and `get` instead of `AtomicCell`. It does make sense to me why there is a desire to avoid dependencies on external crates (`crossbeam`) and to make use of facilities already present in the crate. 

But I am not really sure how to do this, for a couple of reasons:
* I am not sure which `AudioNode` object to employ for this purpose. One option would be to use the [newly created `Var`](https://github.com/SamiPerttu/fundsp/pull/16), although with [possible changes to its API](https://github.com/SamiPerttu/fundsp/pull/18) I would want to wait until that stabilizes before using it.
* `AudioNode::set` and `get`, from what I can tell, operate exclusively on `f64` values. While I could imagine using those for my purposes (e.g., a positive value signals "continue the envelope", a negative value signals "start the release"), I find it a bit inelegant in comparison to a purpose-build `enum`. But if that is what is needed to reduce the external crate dependency, I can work with it.